esr_key_value_database_intrinsics.md

```diff
@@ -0,0 +1,291 @@
+transaction if `itr` wasn't returned by `kv_it_create` or was destroyed.
+
+If `itr`'s status is `iterator_ok` or `iterator_erased`, then this finds the
+next non-deleted key in range. If found, the new status is `iterator_ok`. If
+not found, the new status is `iterator_end`.
+
+If `itr`'s status is `iterator_end` then this finds the first non-deleted key
+in range. If found, the new status is `iterator_ok`. If not found, the new
+status is `iterator_end`.
+
+`kv_it_next` never returns `iterator_erased`.
+
+### kv_it_prev
+
+```c++
+it_stat  kv_it_prev(uint32_t itr);
+```
+
+Move iterator to previous position and return its new status. It aborts the
+transaction if `itr` wasn't returned by `kv_it_create` or was destroyed.
+
+If `itr`'s status is `iterator_ok` or `iterator_erased`, then this finds the
+previous non-deleted key in range. If found, the new status is `iterator_ok`. If
+not found, the new status is `iterator_end`.
+
+If `itr`'s status is `iterator_end` then this finds the last non-deleted key
+in range. If found, the new status is `iterator_ok`. If not found, the new
+status is `iterator_end`.
+
+`kv_it_prev` never returns `iterator_erased`.
+
+### kv_it_lower_bound
+
+```c++
+it_stat  kv_it_lower_bound(uint32_t itr, const char* key, uint32_t size);
+```
+
+Find the first non-deleted key >= the provided key and return the new iterator status.
+It aborts the transaction if `itr` wasn't returned by `kv_it_create` or was destroyed.
+
+If a key is found, the new status is `iterator_ok`. If not found, the new
+status is `iterator_end`. `kv_it_lower_bound` never returns `iterator_erased`.
+
+### kv_it_key
+
+```c++
+it_stat  kv_it_key(uint32_t itr, uint32_t offset, char* dest, uint32_t size, uint32_t& actual_size);
+```
+
+Fetch the key from the iterator and return the iterator's status. It aborts the transaction if
+`itr` wasn't returned by `kv_it_create` or was destroyed.
+
+This sets `actual_size` to the size of the key and copies up to `size` bytes into `dest`.
+
+If `itr`'s status is `iterator_erased`, then this function behaves as if the key was never erased.
+If `itr`'s status is `iterator_end`, then this function behaves as if the key is empty.
+
+`kv_it_key` doesn't modify `dest` if `offset` is greater than the iterator's key size.
+It still aborts the transaction if `[dest, dest + size)` is out of the WASM's linear
+memory range.
+
+WASM memory order: `kv_it_key` writes to `actual_size` after it finishes writing to `dest`.
+
+### kv_it_value
+
+```c++
+it_stat  kv_it_value(uint32_t itr, uint32_t offset, char* dest, uint32_t size, uint32_t& actual_size);
+```
+
+Fetch the value from the iterator and return the iterator's status. It aborts the transaction if
+`itr` wasn't returned by `kv_it_create` or was destroyed.
+
+This sets `actual_size` to the size of the value and copies up to `size` bytes into `dest`.
+
+If `itr`'s status is `iterator_erased` or `iterator_end`, then this function behaves as if the key is empty.
```

Comment:
Did you mean "... as if the **value** is empty"

Reply:
yes

Reply:
fixed